File: frontend/src/contracts/knowledgeContracts.ts
export type KnowledgeDocument = {
  document_id: string;
  title: string;
  doc_type: string;
  content: string;
  service: string | null;
  tags: string[];
  source: string | null;
  archived: boolean;
  metadata: Record<string, unknown>;
  created_at: string | null;
  updated_at: string | null;
  indexed_at: string | null;
  chunk_count: number;
  chunk_ids: string[];
};

export type KnowledgeChunk = {
  chunk_id: string;
  document_id: string;
  title: string;
  doc_type: string;
  text: string;
  chunk_index: number;
  service: string | null;
  tags: string[];
  source: string | null;
  metadata: Record<string, unknown>;
};

export type KnowledgeDocumentDetail = {
  document: KnowledgeDocument;
  chunks: KnowledgeChunk[];
};

export type KnowledgeSearchResult = {
  document_id: string;
  title: string;
  doc_type: string;
  text: string;
  score: number;
  service: string | null;
  tags: string[];
  source: string | null;
  chunk_index: number;
  metadata: Record<string, unknown>;
};

export type KnowledgeDocumentCreatePayload = {
  document_id: string;
  title: string;
  doc_type: string;
  content: string;
  service?: string | null;
  tags?: string[];
  source?: string | null;
  metadata?: Record<string, unknown>;
};

export type KnowledgeDocumentUpdatePayload = {
  title?: string;
  doc_type?: string;
  content?: string;
  service?: string | null;
  tags?: string[];
  source?: string | null;
  archived?: boolean;
  metadata?: Record<string, unknown>;
};

export type KnowledgeSearchPayload = {
  query: string;
  top_k?: number;
  service_name?: string | null;
  doc_types?: string[] | null;
  tags?: string[] | null;
};

File: shared/contracts/knowledge_contracts.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


@dataclass(frozen=True)
class KnowledgeDocument:
    document_id: str
    title: str
    doc_type: str
    content: str
    service: str | None = None
    tags: list[str] = field(default_factory=list)
    source: str | None = None
    archived: bool = False
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "KnowledgeDocument":
        return cls(
            document_id=str(payload["document_id"]),
            title=str(payload["title"]),
            doc_type=str(payload["doc_type"]),
            content=str(payload.get("content", "")),
            service=payload.get("service"),
            tags=list(payload.get("tags", [])),
            source=payload.get("source"),
            archived=bool(payload.get("archived", False)),
            metadata=dict(payload.get("metadata", {})),
        )


@dataclass(frozen=True)
class KnowledgeRetrievalResult:
    document_id: str
    title: str
    doc_type: str
    text: str
    score: float
    service: str | None = None
    tags: list[str] = field(default_factory=list)
    source: str | None = None
    chunk_index: int = 0
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "KnowledgeRetrievalResult":
        return cls(
            document_id=str(payload["document_id"]),
            title=str(payload["title"]),
            doc_type=str(payload["doc_type"]),
            text=str(payload.get("text", "")),
            score=float(payload.get("score", 0.0)),
            service=payload.get("service"),
            tags=list(payload.get("tags", [])),
            source=payload.get("source"),
            chunk_index=int(payload.get("chunk_index", 0)),
            metadata=dict(payload.get("metadata", {})),
        )


@dataclass(frozen=True)
class KnowledgeChunk:
    chunk_id: str
    document_id: str
    title: str
    doc_type: str
    text: str
    chunk_index: int
    service: str | None = None
    tags: list[str] = field(default_factory=list)
    source: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "KnowledgeChunk":
        return cls(
            chunk_id=str(payload["chunk_id"]),
            document_id=str(payload["document_id"]),
            title=str(payload["title"]),
            doc_type=str(payload["doc_type"]),
            text=str(payload.get("text", "")),
            chunk_index=int(payload.get("chunk_index", 0)),
            service=payload.get("service"),
            tags=list(payload.get("tags", [])),
            source=payload.get("source"),
            metadata=dict(payload.get("metadata", {})),
        )
